File: server/lib/dark_worlds_server/engine/projectile.ex
defmodule DarkWorldsServer.Engine.Projectile do
  use DarkWorldsServer.Communication.Encoder

  @enforce_keys [
    :id,
    :position,
    :direction,
    :speed,
    :range,
    :player_id,
    :damage,
    :remaining_ticks,
    :projectile_type,
    :status,
    :last_attacked_player_id,
    :pierce,
    :projectile_model_name
  ]
  defstruct [
    :id,
    :position,
    :direction,
    :speed,
    :range,
    :player_id,
    :damage,
    :remaining_ticks,
    :projectile_type,
    :status,
    :last_attacked_player_id,
    :pierce,
    :projectile_model_name
  ]
end

File: client/Assets/Scripts/Game/Projectile.cs
namespace Game {
    using Communication.Protobuf;

    public class Projectile {
        private ulong id { get; }
        private string name { get; }
        private ulong damage { get; }
        private float speed { get; }
        private float size { get; }
        private Position position { get; }
        private float direction { get; }

        public Projectile(Communication.Protobuf.Projectile protobufProjectile) {
            this.id = protobufProjectile.Id;
            this.name = protobufProjectile.Name;
            this.damage = protobufProjectile.Damage;
            this.speed = protobufProjectile.Speed;
            this.size = protobufProjectile.Size;
            this.direction = protobufProjectile.Direction;
            this.position = new Position(protobufProjectile.Position);
        }
    }
}
